File: solutions/javascript/two-bucket/1/two-bucket.js

const gcd = (a, b) => (b ? gcd(b, a % b) : a)

export class TwoBucket {
  buckets = [];
  goalAmt = 0;
  moves = 0;

  constructor(b1Amt, b2Amt, goal, startBucket) {
    this.goalAmt = goal;
    this.buckets = [
      {name: 'one', size: b1Amt, current: 0},
      {name: 'two', size: b2Amt, current: 0}];
    if ( startBucket === 'two' ) {
      this.buckets.reverse();
    }
    this.#assertIsValid();
  }

  solve() {
    let moves = 0

    while(this.buckets[0].current !== this.goalAmt && this.buckets[1].current !== this.goalAmt) {
      moves += 1

      if(this.buckets[0].current === 0) this.buckets[0].current = this.buckets[0].size
      else if (this.buckets[1].current === this.buckets[1].size) this.buckets[1].current = 0
      else if (this.buckets[1].size === this.goalAmt) this.buckets[1].current = this.buckets[1].size
      else {
        const toTransfer = Math.min(this.buckets[1].size - this.buckets[1].current, this.buckets[0].current)
        this.buckets[0].current -= toTransfer
        this.buckets[1].current += toTransfer
      }
    }

    this.moves = moves;
    return this;
  }

  get goalBucket() {
    return this.buckets[0].current === this.goalAmt ? this.buckets[0].name : this.buckets[1].name;
  }

  get otherBucket() {
    return this.buckets[0].current === this.goalAmt ? this.buckets[1].current : this.buckets[0].current;
  }

  #assertIsValid() {
    if(this.goalAmt > this.buckets[0].size && this.goalAmt > this.buckets[1].size) throw new Error('goal unreachable')

    if (this.goalAmt % gcd(this.buckets[0].size, this.buckets[1].size) !== 0) {
      throw new Error('goal unreachable')
    }
  }
}

File: solutions/ruby/two-bucket/1/two_bucket.rb
class TwoBucket
    def initialize(bucket_one, bucket_two, goal, start_bucket)
        @buckets = [{name: "one", size: bucket_one, current: 0}, {name: "two", size: bucket_two, current: 0}]
        @buckets = @buckets.reverse if start_bucket == "two"
        @goal = goal
        @start_bucket = start_bucket
        @moves = 0
        solve
    end

    def solve
        solved_bucket = nil
        until @buckets.find { |b| b[:current] == @goal } != nil
            @moves += 1
            first, second = @buckets
            if first[:current].zero?
                first[:current] = first[:size]
            elsif second[:size] == @goal
                second[:current] = second[:size]
            elsif second[:current] < second[:size]
                first, second = transfer(first, second)
            else
                second[:current] = 0
            end
            @buckets = [first, second]
        end
    end

    def moves
        @moves
    end

    def goal_bucket
        @buckets.find { |b| b[:current] == @goal }[:name]
    end

    def other_bucket
        @buckets.find { |b| b[:current] != @goal }[:current]
    end

    #private
    def transfer(from, to)
        if from[:current] <= to[:size] - to[:current]
            to[:current] += from[:current]
            from[:current] = 0
        else
            from[:current] -= to[:size] - to[:current]
            to[:current] = to[:size]
        end
        [from, to]
    end
end